File: epi_judge_java/epi/IsTreeBalanced.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class IsTreeBalanced {

    private static int findHeightDfs(BinaryTreeNode<Integer> tree) {
        if (tree == null) return 0;
        int left = findHeightDfs(tree.left);
        if (left == -1) return -1;
        int right = findHeightDfs(tree.right);
        if (right == -1) return -1;
        if (Math.abs(left - right) > 1) return -1;
        return Math.max(left, right) + 1;
    }

    @EpiTest(testDataFile = "is_tree_balanced.tsv")
    public static boolean isBalanced(BinaryTreeNode<Integer> tree) {
        return findHeightDfs(tree) != -1;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "IsTreeBalanced.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/is_tree_balanced.cc
#include "binary_tree_node.h"
#include "test_framework/generic_test.h"

using namespace std;

int height(const unique_ptr<BinaryTreeNode<int>> &tree) {
    if (tree == nullptr) return 0;
    return max(height(tree->left), height(tree->right)) + 1;
}

bool IsBalanced(const unique_ptr<BinaryTreeNode<int>> &tree) {
    if (tree == nullptr) return true;
    if (abs(height(tree->left) - height(tree->right)) > 1) return false;
    return IsBalanced(tree->left) && IsBalanced(tree->right);
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"tree"};
    return GenericTestMain(args, "is_tree_balanced.cc", "is_tree_balanced.tsv",
                           &IsBalanced, DefaultComparator{}, param_names);
}
